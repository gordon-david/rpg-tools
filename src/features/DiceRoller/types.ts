export interface Dice {
    id: string
}