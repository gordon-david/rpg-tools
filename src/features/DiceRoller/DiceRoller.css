.dice-roller {
    max-width: 320px;
    margin: 1rem;
    border: 1px solid gold;
}

.dice-roller__dice-grid {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
}

.dice {
    width: 5rem;
    display: flex;
    flex-direction: column;
}

.dice__amount {
    text-align: center;
}

.dice-roller__controls {
    display: flex;
    justify-content: space-around;
}
