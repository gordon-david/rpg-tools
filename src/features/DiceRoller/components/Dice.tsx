export interface DiceProps {
    displayName: string,
    diceType: string,
    amount: number,
    increment: () => void
}

/*
   Displays a dice button,
   with an amount and displayname.
   Accepts a callback for incrementing the amount
   of dice displayed.
*/
export function Dice({
    displayName,
    diceType,
    amount,
    increment }: DiceProps) {

    return (
        <div
            className="dice"
            id={diceType}>
            <button
                className='increment'
                onClick={increment}
            >
                {displayName}
            </button>
            <div className='amount'>
                {amount}
            </div>
        </div>
    );
}
