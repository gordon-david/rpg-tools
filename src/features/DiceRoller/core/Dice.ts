enum DiceKey {
    D4 = 'polyhedral-d4',
    D6 = 'polyhedral-d6',
    D8 = 'polyhedral-d8',
    D10 = 'polyhedral-d10',
    D12 = 'polyhedral-d12',
    D20 = 'polyhedral-d20',
    D100 = 'polyhedral-d100'
}

class PolyehedralDice {
    max: number
    key: DiceKey
    displayName: string

    constructor(displayName: string, key: DiceKey, max: number) {
        this.displayName = displayName
        this.key = key,
            this.max = max
    }

    roll(): number {
        return Math.floor(Math.random() * this.max) + 1
    }
}

class DiceFactory {
    static build(key: DiceKey): PolyehedralDice {
        switch (key) {
            case DiceKey.D4: return new PolyehedralDice('D4', key, 4)
            case DiceKey.D6: return new PolyehedralDice('D6', key, 6)
            case DiceKey.D8: return new PolyehedralDice('D8', key, 8)
            case DiceKey.D10: return new PolyehedralDice('D10', key, 10)
            case DiceKey.D12: return new PolyehedralDice('D12', key, 12)
            case DiceKey.D20: return new PolyehedralDice('D20', key, 20)
            case DiceKey.D100: return new PolyehedralDice('D100', key, 100)
            default: throw new Error("Dice type unsupported.")
        }
    }
}

class DicePool {
    private diceMap: { [DiceKey: string]: number }
    private diceCache: { [DiceKey: string]: PolyehedralDice }

    constructor(dicekeys: DiceKey[]) {
        this.diceMap = {}
        this.diceCache = {}
        console.log('new');
        

        dicekeys.forEach(key => {
            this.diceMap[key] = 0
            this.diceCache[key] = DiceFactory.build(key)
        })
    }

    add(diceKey: DiceKey) { this.diceMap[diceKey] += 1 }
    remove(diceKey: DiceKey) {
        if (this.diceMap[diceKey] < 1) return
        this.diceMap[diceKey] -= 1
    }

    getDiceKeys() { return Object.keys(this.diceMap) }
    getDiceMap() { return this.diceMap }

    roll(): number {
        let result = 0

        Object.keys(this.diceMap).forEach(key => {
            for (let rolls = 0; rolls < this.diceMap[key]; rolls++) {
                result += this.diceCache[key].roll()
            }
        })

        return result
    }

}

function PolyhedralDicePoolFactory(): DicePool {

    return new DicePool([
        DiceKey.D4,
        DiceKey.D6,
        DiceKey.D8,
        DiceKey.D10,
        DiceKey.D12,
        DiceKey.D20,
        DiceKey.D100,
    ])

}

export { PolyhedralDicePoolFactory, DicePool, DiceKey }