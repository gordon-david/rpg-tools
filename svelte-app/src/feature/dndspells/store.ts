import { derived, get, writable, type Readable, type Subscriber, type Writable } from "svelte/store"

import data from '$lib/json/spells.json'
import { readable } from "svelte/store"
import type { Filter, Sorting, Spell, SpellData, TableColumn } from "./types";

const initialData: any = {
    spells: [],
    filters: []
}

export interface SpellStore {
    spells: Readable<Spell[]>,
    schools: string[],
    classes: string[],
    backgrounds: string[],
    toggleSortDirection: (by?: TableColumn) => void,
    setFilters: (filters: Filter[]) => void,
    setSorting: (by: TableColumn) => void
}

type SpellWrapper = {
    index: number,
    spell: Spell,
    included: boolean
}

export function makeSpellStore(): SpellStore {

    const sorting: Sorting = { by: 'name', direction: 'ascending' }
    const spellWrapper: Writable<SpellWrapper[]> = writable((data as SpellData)['spells']
        .map((spell: Spell, index: number) => ({
            spell,
            index,
            included: true
        })))

    const spells: Readable<Spell[]> = derived(spellWrapper, (values: SpellWrapper[]) => values.reduce(
        (result: Spell[], element: SpellWrapper) => {
            if (element.included)
                result.push(element.spell)
            return result
        },
        []
    ))


    function applySorting() {
        spellWrapper.update((value: SpellWrapper[]) => {
            value.sort((prev: SpellWrapper, next: SpellWrapper) => {

                let prevSortVal = ''
                let nextSortVal = ''

                switch (sorting.by) {
                    case 'name':
                        prevSortVal = prev.spell.name
                        nextSortVal = next.spell.name
                        break
                    case 'school':
                        prevSortVal = prev.spell.school
                        nextSortVal = next.spell.school
                        break
                    default:
                        return 0
                }
                if (sorting.direction === 'ascending') {
                    return (prevSortVal < nextSortVal) ? -1 : 1
                }
                return (prevSortVal > nextSortVal) ? -1 : 1
            })

            console.log([value[0].spell.school]);
            return value
        })
    }


    return {
        classes: data.classes,
        schools: data.schools,
        backgrounds: data.backgrounds,
        spells: { subscribe: spells.subscribe },
        toggleSortDirection(by?: TableColumn) {
        },
        setFilters(filters: Filter[]) {
            console.log(filters);
            if (filters.length > 0) {

                spellWrapper.update((spellWrapper: SpellWrapper[]) =>
                    spellWrapper.map((spellWrapper) => {
                        let include = false
                        let filterIndex = 0

                        while (!include && filterIndex < filters.length) {
                            const currentFilter = filters[filterIndex]

                            switch (filters[filterIndex].type) {
                                case 'school':
                                    // avoiding assignment on short circuit
                                    if (currentFilter.value.toLowerCase() === spellWrapper.spell.school.toLowerCase()) {
                                        console.log(spellWrapper.spell.school);
                                        include = true
                                    }
                                    break
                                case "class":
                                    spellWrapper.spell.classes.fromClassList.forEach(
                                        classElement => {
                                            if (classElement.name.toLowerCase() === currentFilter.value.toLowerCase()){
                                                include = true
                                            }

                                        }
                                    )
                                    if (spellWrapper.spell.classes.fromClassListVariant !== undefined) {
                                        spellWrapper.spell.classes.fromClassListVariant.forEach(
                                            classElement => {
                                                if (classElement.name.toLowerCase() === currentFilter.value.toLowerCase())
                                                    include = true
                                            }
                                        )
                                    }
                            }

                            if (include)
                                break

                            filterIndex += 1
                        }

                        spellWrapper.included = include
                        return spellWrapper
                    })
                )
                return
            }
            spellWrapper.update(spellWrapper => spellWrapper.map(sw => {
                sw.included = true
                return sw
            }))
        },
        setSorting(by: TableColumn) {
            if (sorting.by === by) {
                if (sorting.direction === 'ascending') {
                    sorting.direction = 'descending'
                } else {
                    sorting.direction = 'ascending'
                }
            } else {
                console.log('here');

                sorting.by = by
                sorting.direction = 'ascending'
            }

            applySorting()
        }
    }
}