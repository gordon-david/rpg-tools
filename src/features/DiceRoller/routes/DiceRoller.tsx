import { useEffect, useState } from "react";
import { Dice } from "../components/Dice";
import { DicePool } from "../components/DicePool";
import { DiceKey, PolyhedralDicePoolFactory } from "../core/Dice";


export function DiceRoller() {
    const [dicePool, setDicePool] = useState(PolyhedralDicePoolFactory())
    const [results, setResults] = useState([])
    const [diceMap, setDiceMap] = useState({})

    useEffect(() => {
        setDiceMap(dicePool.getDiceMap)
    }, [dicePool])

    function addDice(key: DiceKey){
        // add to dicePool
        dicePool.add(key)
        setDicePool(dicePool)
    }

    function removeDice(key: DiceKey){
        dicePool.remove(key)
        setDicePool(dicePool)
    }

    return <>

        <div className="dice-roller">
            <DicePool />
            <div>Roll Results List</div>
        </div>
    </>
}

