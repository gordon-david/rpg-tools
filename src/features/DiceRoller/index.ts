import { DiceRoller } from "./routes/DiceRoller";
export * from "./types"

export {DiceRoller}