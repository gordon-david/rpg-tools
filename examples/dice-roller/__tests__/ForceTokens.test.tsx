import React from 'react'
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import { ForceTokens } from '../components/DiceRoller/ForceTokens'
import { shallow, mount } from 'enzyme'
import { reducers, actionCreators } from '../store'
import { DiceFaces } from '../utilities/Dice'

test('ForceTokens component renders', () => {
    const wrapper = shallow(<Provider store={createStore(reducers)}><ForceTokens /></Provider>)
    expect(wrapper).toMatchSnapshot()
})

test('Initial store state is filtered for force dice roll results', () => {
    const store = createStore(reducers)
    store.dispatch(actionCreators.addRollResult([DiceFaces.Genesys.ADVANTAGE]))
    store.dispatch(actionCreators.addRollResult([DiceFaces.Genesys.ADVANTAGE, DiceFaces.Genesys.LIGHT]))
    store.dispatch(actionCreators.addRollResult([DiceFaces.Genesys.LIGHT, DiceFaces.Genesys.DARK]))

    const wrapper = mount(<Provider store={store}><ForceTokens /></Provider>)
    expect(wrapper.find('#token-display').children().length).toBe(3)
})

test('as roll results are added to the store, new force die results are added', () => {
    const store = createStore(reducers)
    const wrapper = mount(<Provider store={store}><ForceTokens /></Provider>)
    expect(wrapper.find('#token-display').children().length).toBe(0)
    store.dispatch(actionCreators.addRollResult([DiceFaces.Genesys.ADVANTAGE, DiceFaces.Genesys.LIGHT]))
    wrapper.setProps({ rollResults: store.getState().rollResults })
    expect(wrapper.find('#token-display').children().length).toBe(1)
})

test('increment button adds a force token', () => {
    const wrapper = mount(<Provider store={createStore(reducers)}><ForceTokens/></Provider>)
    // assert default number of force tokens
    const startNumTokens = wrapper.find('#token-display').children().length
    // event click inrement-force-token button
    wrapper.find('#increment').simulate('click')
    // assert increase of force tokens
    expect(wrapper.find('#token-display').children().length).toBe(startNumTokens + 1)
})

test('decrement button removes a force token', () => {
    const wrapper = mount(<Provider store={createStore(reducers)}><ForceTokens/></Provider>)
    wrapper.find('#increment').simulate('click')
    const startNumTokens = wrapper.find('#token-display').children().length
    wrapper.find('#decrement').simulate('click')
    expect(wrapper.find('#token-display').children().length).toBe(startNumTokens - 1)
})

test('clicking a force token toggles light and dark', () => {
    const wrapper = mount(<Provider store={createStore(reducers)}><ForceTokens/></Provider>)
    wrapper.find('#increment').simulate('click')
    expect(wrapper.exists(`.${DiceFaces.Genesys.LIGHT}`)).toBe(true)
    expect(wrapper.exists(`.${DiceFaces.Genesys.DARK}`)).toBe(false)
    wrapper.find(`.${DiceFaces.Genesys.LIGHT}`).simulate('click')
    expect(wrapper.exists(`.${DiceFaces.Genesys.LIGHT}`)).toBe(false)
    expect(wrapper.exists(`.${DiceFaces.Genesys.DARK}`)).toBe(true)

})
