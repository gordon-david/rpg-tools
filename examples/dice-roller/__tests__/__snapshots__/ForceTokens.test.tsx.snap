// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForceTokens component renders 1`] = `ShallowWrapper {}`;
